lms/services/blackboard_api.py

```diff
@@ -1,6 +1,32 @@
+from marshmallow import INCLUDE
+
+from lms.services.exceptions import HTTPError, OAuth2TokenError
+from lms.validation import RequestsResponseSchema, ValidationError
 from lms.validation.authentication import OAuthTokenResponseSchema
 
 
+class BlackboardErrorResponseSchema(RequestsResponseSchema):
+    """
+    Schema for validating error responses from the Blackboard API.
+
+    This schema's parse() method will return the given response's JSON body as
+    a dict (with whatever fields the response body from Blackboard had) and
+    will default to returning an empty dict if the given response's body isn't
+    a JSON dict, or isn't a dict at all, or if the response isn't a
+    requests.Response object (e.g. if it's None).
+
+    """
+
+    class Meta:
+        unknown = INCLUDE
+
+    def parse(self, *args, **kwargs):
+        try:
+            return super().parse(*args, **kwargs)
+        except ValidationError:
+            return {}
```

Comment:
This schema handles quite a lot for us in very little code. You can just do:

```python
error_dict = BlackboardErrorResponseSchema(err.response).parse()
```

and can then go ahead and assume that `error_dict` is a dict (although you can't assume anything about the contents of the dict) even though `err.response` could have been:

* `None` (or anything else that doesn't have a `.json()` method like `requests.Response` does)
* A `requests.Response` object with a non-JSON body
* A response with a non-dict JSON body (e.g. a list)

And the schema can be re-used wherever we want to parse Blackboard error responses in future.

It could possibly be rewritten without Marshmallow. Something like:

```python
def error_dict(response):
    try:
        error_dict = response.json()
    except (AttributeError, ValueError):
        error_dict = {}

    if not isinstance(error_dict, dict):
        error_dict = {}

    return error_dict
```

... but this would kind of just be duplicating code that Marshmallow and `RequestsResponseSchema` already contains, so I'm not sure.